### PR TITLE
[show][config] fix show mux status health field; add support for hwmode functionality to toggle mux, check mux direction for Y cable

### PR DESCRIPTION
**Pull request by vdahiya12**
#### Summary

This PR fixes the show mux status  "health" field. 

There is also support for new commands for muxcable
show muxcable hwmode muxdirection 
sudo config muxcable hwmode state active all
which can check the muxdirection and toggle the mux direction directly through eeprom

#### What I did
Fixed the show mux status health field
Also added these new commands

```
vdahiya@BN9-0101-0301-01UT0:~$ show muxcable hwmode muxdirection 
Port         Direction
-----------  -----------
Ethernet0    active
Ethernet4    active
Ethernet8    active
Ethernet12   active
Ethernet16   active
```

```
vdahiya@BN9-0101-0301-01UT0:~$ sudo config muxcable hwmode state active all
Muxcables at all ports will be changed to active state. Continue? [y/N]: y
Success in toggling port Ethernet0 to active
Success in toggling port Ethernet4 to active
Success in toggling port Ethernet8 to active
Success in toggling port Ethernet12 to active
Success in toggling port Ethernet16 to active
```

#### How I did it
Added the changes to show/muxcable.py to correctly retrieve health "state"  field from the "MUX_LINKMGR_TABLE"  state DB table.
For new cli commands added the support in show/muxcable.py and config/muxcable.py

#### How to verify it
Ran the command on the 7050 testbed

#### New command output (if the output of a command-line utility has changed)

Signed-off-by: vaibhav-dahiya <vdahiya@microsoft.com>

### CODE REVIEW

File: show/muxcable.py [hunk 1]
```diff
@@ -155,22 +156,25 @@ def status(port, json_output):
 
         muxcable_info_dict[asic_index] = per_npu_statedb[asic_index].get_all(
             per_npu_statedb[asic_index].STATE_DB, 'MUX_CABLE_TABLE|{}'.format(port))
+        muxcable_health_dict[asic_index] = per_npu_statedb[asic_index].get_all(
+            per_npu_statedb[asic_index].STATE_DB, 'MUX_LINKMGR_TABLE|{}'.format(port))
         if muxcable_info_dict[asic_index] is not None:
             logical_key = "MUX_CABLE_TABLE"+"|"+port
+            logical_health_key = "MUX_LINKMGR_TABLE"+"|"+port
```
Comment (jleveque) [inline show/muxcable.py:163]: LGTM alerts that `logical_health_key` is unused.
Also, please add spaces around `+` signs.
Furthermore, suggest `logical_key = "MUX_CABLE_TABLE|" + port` or `logical_key = "MUX_CABLE_TABLE|{}".format(port)`

**Review by jleveque — CHANGES_REQUESTED** [review]

File: config/muxcable.py [hunk 2]
```diff
@@ -257,3 +262,107 @@ def disable(port, target):
...
+        click.confirm(('Do you wish to change all the ports to {} state Continue?'.format(state)), abort=True)
+        logical_port_list = platform_sfputil.logical
+
+        rc = True
+        for port in logical_port_list:
+            if platform_sfputil is not None:
+                physical_port_list = platform_sfputil_helper.logical_port_name_to_physical_port_list(port)
+
+            if not isinstance(physical_port_list, list):
+                click.echo("ERR: Unable to get a port on muxcable port")
```
Comment (jleveque) [inline config/muxcable.py:331]: Message isn't very clear. Maybe `"ERR: Unable to locate physical port information for {}".format(port)`?

File: config/muxcable.py [hunk 3]
```diff
@@ -257,3 +262,107 @@ def disable(port, target):
...
+                rc = False
+                if len(physical_port_list) != 1:
+                    click.echo("ERR: Unable to get a single port on muxcable")
```
Comment (jleveque) [inline config/muxcable.py:334]: Message isn't very clear. Maybe `"ERR: Found multiple physical ports associated with {}".format(port)`? And maybe even print out the list of physical ports to aid in debugging?

File: show/muxcable.py [hunk 4]
```diff
@@ -419,3 +427,97 @@ def cableinfo(port):
...
+@click.argument('port', metavar='<port_name>', required=False, default=None)
+def mux_direction(port):
+    if port is not None:
+
+        """Show muxcable summary information"""
+        if platform_sfputil is not None:
+            physical_port_list = platform_sfputil_helper.logical_port_name_to_physical_port_list(port)
+
+        if not isinstance(physical_port_list, list):
+            click.echo("ERR: Unable to get a port on muxcable port")
```
Comment (jleveque) [inline show/muxcable.py:448]: Message isn't very clear. Maybe `"ERR: Unable to locate physical port information for {}".format(port)`?

File: show/muxcable.py [hunk 5]
```diff
@@ -419,3 +427,97 @@ def cableinfo(port):
...
+            sys.exit(EXIT_FAIL)
+            if len(physical_port_list) != 1:
+                click.echo("ERR: Unable to get a single port on muxcable")
```
Comment (jleveque) [inline show/muxcable.py:451]: Message isn't very clear. Maybe `"ERR: Found multiple physical ports associated with {}".format(port)`? And maybe even print out the list of physical ports to aid in debugging?

File: show/muxcable.py [hunk 6]
```diff
@@ -419,3 +427,97 @@ def cableinfo(port):
...
+                    state = "standby"
+                click.echo("mux direction for the cable at port {} is {}".format(port, state))
+            elif int(read_side) == 2:
+                if mux_direction == 1:
+                    state = "standby"
+                elif mux_direction == 2:
+                    state = "active"
+                click.echo("mux direction for the cable at port {} is {}".format(port, state))
+            else:
+                click.echo("ERR: Unable to get mux direction, direction in unknown")
```
Comment (jleveque) [inline show/muxcable.py:523]: s/in/is/

File: show/muxcable.py [hunk 7]
```diff
@@ -419,3 +427,97 @@ def cableinfo(port):
...
+                state = "standby"
+            click.echo("mux direction for the cable at port {} is {}".format(port, state))
+        elif int(read_side) == 2:
+            if mux_direction == 1:
+                state = "standby"
+            elif mux_direction == 2:
+                state = "active"
+            click.echo("mux direction for the cable at port {} is {}".format(port, state))
+        else:
+            click.echo("ERR: Unable to get mux direction, direction in unknown")
```
Comment (jleveque) [inline show/muxcable.py:479]: s/in/is/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: show/muxcable.py [hunk 8]
```diff
@@ -412,3 +427,97 @@ def cableinfo(port):
 
     body = [[vendor, part_num]]
     click.echo(tabulate(body, headers=headers))
+
+
+@muxcable.group(cls=clicommon.AbbreviationGroup)
+def hwmode():
+    """start configuring cable in hardware mode on a port"""
+    pass
```
Comment (tahmed-dev) [inline show/muxcable.py:435]: should we be printing some error message related to missing params.

File: show/muxcable.py [hunk 9]
```diff
@@ -412,3 +427,97 @@ def cableinfo(port):
...
+
+        if not isinstance(physical_port_list, list):
+            click.echo("ERR: Unable to get a port on muxcable port")
+            sys.exit(EXIT_FAIL)
+            if len(physical_port_list) != 1:
+                click.echo("ERR: Unable to get a single port on muxcable")
+                sys.exit(EXIT_FAIL)
+
+        physical_port = physical_port_list[0]
+        import sonic_y_cable.y_cable
```
Comment (tahmed-dev) [inline show/muxcable.py:464]: Why import inside a method and not at the top?

File: tests/mock_tables/state_db.json [hunk 10]
```diff
@@ -490,6 +490,21 @@
     "MUX_CABLE_TABLE|Ethernet12": {
         "state": "unknown"
     },
+    "MUX_LINKMGR_TABLE|Ethernet32": {
+        "state": "healthy"
+    },
+    "MUX_LINKMGR_TABLE|Ethernet0": {
+        "state": "healthy"
+    },
+    "MUX_LINKMGR_TABLE|Ethernet4": {
```
Comment (tahmed-dev) [inline tests/mock_tables/state_db.json:499]: can we mark few as unhealthy?

File: config/muxcable.py [hunk 11]
```diff
@@ -276,28 +276,34 @@ def hwmode():
 def state(state, port):
     """Show muxcable summary information"""
     if port is not None and port != "all":
-        click.confirm(('Do you wish to change the port {} to {} state, Continue?'.format(port, state)), abort=True)
+        click.confirm(('Do you wish to change the muxcable port {} to {} state, Continue?'.format(port, state)), abort=True)
```
Comment (jleveque) [inline config/muxcable.py:279]: Suggest rephrasing message: `'Muxcable at port {} will be changed to {} state. Continue?'

**Review by jleveque — CHANGES_REQUESTED** [review]

File: show/muxcable.py (hunk 1 above)
Comment (vdahiya12) [reply show/muxcable.py]: fixed

File: config/muxcable.py (hunk 2 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: config/muxcable.py (hunk 3 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: show/muxcable.py (hunk 4 above)
Comment (vdahiya12) [reply show/muxcable.py]: fixed

File: show/muxcable.py (hunk 5 above)
Comment (vdahiya12) [reply show/muxcable.py]: fixed

File: show/muxcable.py (hunk 6 above)
Comment (vdahiya12) [reply show/muxcable.py]: changed the log message, fixed

File: show/muxcable.py (hunk 7 above)
Comment (vdahiya12) [reply show/muxcable.py]: changed the log message, fixed

File: show/muxcable.py (hunk 9 above)
Comment (vdahiya12) [reply show/muxcable.py:464]: importing at top might be redundant, because some commands
like show mux status/config do not need sonic_y_cable package.
Thats why prefer to have it inside the command.

File: show/muxcable.py (hunk 8 above)
Comment (vdahiya12) [reply show/muxcable.py:435]: I think click does print missing params by default. 

```admin@STR43-0101-0101-01LT0:~$ show mux hwmode 
Usage: show mux hwmode [OPTIONS] COMMAND [ARGS]...

  start configuring cable in hardware mode on a port

Options:
  -h, -?, --help  Show this message and exit.

Commands:
  mux-direction  Show muxcable mux_direction information```

File: config/muxcable.py (hunk 11 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: config/muxcable.py [hunk 12]
```diff
@@ -322,12 +322,12 @@ def state(state, port):
...
 
         if res == False:
-            click.echo("ERR: Unable to toggle port to {}".format(port, state))
+            click.echo("ERR: Unable to toggle port {} to {}".format(port, state))
             sys.exit(CONFIG_FAIL)
 
     elif port == "all" and port is not None:
 
-        click.confirm(('Do you wish to change all the muxcable ports to {} state Continue?'.format(state)), abort=True)
+        click.confirm(('Muxcable at all ports will be changed to {} state. Continue?'.format(state)), abort=True)
```
Comment (jleveque) [inline config/muxcable.py:330]: s/Muxcable/Muxcables/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/mock_tables/state_db.json (hunk 10 above)
Comment (vdahiya12) [reply tests/mock_tables/state_db.json:499]: fixed

File: config/muxcable.py [hunk 13]
```diff
@@ -344,17 +413,44 @@ def state(state, port):
...
+            transceiver_dict[asic_index] = per_npu_statedb[asic_index].get_all(
+                per_npu_statedb[asic_index].STATE_DB, 'TRANSCEIVER_INFO|{}'.format(port))
+            vendor_value = transceiver_dict[asic_index].get("manufacturer", None)
+            model_value = transceiver_dict[asic_index].get("model", None)
+
+            """ This check is required for checking whether or not this port is connected to a Y cable
+            or not. The check gives a way to differentiate between non Y cable ports and Y cable ports.
+            TODO this should be removed once their is support for multiple vendors on Y cable"""
+
+            if vendor_value != VENDOR_NAME and model_value != VENDOR_MODEL:
```
Comment (jleveque) [inline config/muxcable.py:425]: I believe this should be `or`, not `and`.

File: show/muxcable.py [hunk 14]
```diff
@@ -448,6 +468,34 @@ def mux_direction(port):
...
+        click.echo("debug {} {} {}".format(asic_index, port, transceiver_dict[asic_index]))
+
+        vendor_value = get_value_for_key_in_dict(transceiver_dict[asic_index], port, "manufacturer", "TRANSCEIVER_INFO")
+        model_value = get_value_for_key_in_dict(transceiver_dict[asic_index], port, "model", "TRANSCEIVER_INFO")
+
+        """ This check is required for checking whether or not this port is connected to a Y cable
+        or not. The check gives a way to differentiate between non Y cable ports and Y cable ports.
+        TODO: this should be removed once their is support for multiple vendors on Y cable"""
+
+        if vendor_value != VENDOR_NAME and model_value != VENDOR_MODEL:
```
Comment (jleveque) [inline show/muxcable.py:495]: I believe this should be `or`, not `and`.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: config/muxcable.py [hunk 15]
```diff
@@ -406,12 +406,13 @@ def state(state, port):
...
-                    click.echo(("ERR: Found multiple physical ports associated with {}, physical Ports:".format(port)))
-                    for lport in physical_port_list:
-                        click.echo(("{}".format(lport)))
-                    rc = False
+                continue
+
+            if len(physical_port_list) != 1:
+                click.echo(("ERR: Found multiple physical ports associated with {}, physical Ports:".format(port)))
+                for lport in physical_port_list:
+                    click.echo(("{}".format(lport)))
```
Comment (jleveque) [inline config/muxcable.py:414]: You can omit the trailing newline by passing `nl=False` (see [here](https://click.palletsprojects.com/en/7.x/utils/)). Also, you have unnecessary double parentheses.

However, I suggest doing something like this instead:

```python
click.echo("ERR: Found multiple physical ports ({}) associated with {}".format(", ".join(physical_port_list), port))
```

**Review by jleveque — CHANGES_REQUESTED** [review]

File: config/muxcable.py (hunk 12 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: config/muxcable.py (hunk 13 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: show/muxcable.py (hunk 14 above)
Comment (vdahiya12) [reply show/muxcable.py]: fixed

File: config/muxcable.py (hunk 15 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

**Comment by vdahiya12** [comment]
Retest this please

File: show/muxcable.py [hunk 16]
```diff
@@ -441,7 +441,7 @@ def hwmode():
 
 @hwmode.command()
 @click.argument('port', metavar='<port_name>', required=False, default=None)
-def mux_direction(port):
+def muxdirection(port):
     """Show muxcable mux_direction information"""
```
Comment (jleveque) [inline show/muxcable.py:445]: Since this docstring becomes the brief help message, please make more readable. Like "Shows the current direction of the cable mux" or similar. Also review others in the file and update if needed.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: show/muxcable.py (hunk 16 above)
Comment (vdahiya12) [reply show/muxcable.py]: fixed

File: show/muxcable.py (hunk 9 above)
Comment (jleveque) [reply show/muxcable.py:521]: >importing at top might be redundant

I think "unnecessary" is more correct than "redundant" here. However, I do prefer importing at the top whenever possible. For now we can leave this as-is and revisit after the y_cable lib refactor.

File: show/muxcable.py (hunk 8 above)
Comment (jleveque) [reply show/muxcable.py:439]: Yes, Click will print the help message for the current command if required params are missing. This is the beauty of using a third-part CLI library :)

File: show/muxcable.py [hunk 17]
```diff
@@ -412,3 +431,214 @@ def cableinfo(port):
...
+@muxcable.group(cls=clicommon.AbbreviationGroup)
+def hwmode():
+    """start configuring cable in hardware mode on a port"""
+    pass
+
+
+@hwmode.command()
+@click.argument('port', metavar='<port_name>', required=False, default=None)
+def muxdirection(port):
+    """Shows the current direction of the cable muxcable {active/standy}"""
```
Comment (jleveque) [inline show/muxcable.py:445]: `cable muxcable` seems redundant.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: show/muxcable.py (hunk 17 above)
Comment (vdahiya12) [reply show/muxcable.py]: fixed

File: show/muxcable.py [hunk 18]
```diff
@@ -125,9 +130,10 @@ def create_json_dump_per_port_config(port_status_dict, per_npu_configdb, asic_id
 def status(port, json_output):
     """Show muxcable status information"""
 
-    port_table_keys = {}
+    port_table_keys, port_health_table_keys = {}, {}
```
Comment (jleveque) [inline show/muxcable.py:133]: Suggest defining these on separate lines

File: config/muxcable.py [hunk 19]
```diff
@@ -146,7 +148,7 @@ def mode(state, port, json_output):
 
         if per_npu_statedb[asic_index] is not None:
             y_cable_asic_table_keys = port_table_keys[asic_index]
-            logical_key = "MUX_CABLE_TABLE"+"|"+port
+            logical_key = "MUX_CABLE_TABLE" + "|" + port
```
Comment (jleveque) [inline config/muxcable.py:151]: Suggest `logical_key = "MUX_CABLE_TABLE|{}".format(port)`

**Review by jleveque — CHANGES_REQUESTED** [review]

File: config/muxcable.py [hunk 20]
```diff
@@ -257,3 +264,208 @@ def disable(port, target):
         sys.exit(CONFIG_FAIL)
     click.echo("loopback disable sucessful")
     sys.exit(CONFIG_SUCCESSFUL)
+
+
+@muxcable.group(cls=clicommon.AbbreviationGroup)
+def hwmode():
+    """start configuring cable in hw mode on a port"""
```
Comment (jleveque) [inline config/muxcable.py:271]: Suggest rewording to `Configure muxcable hardware directly" or similar.

File: config/muxcable.py [hunk 21]
```diff
@@ -257,3 +264,208 @@ def disable(port, target):
...
+    pass
+
+
+@hwmode.command()
+@click.argument('state', metavar='<operation_status>', required=True, type=click.Choice(["active", "standby"]))
+@click.argument('port', metavar='<port_name>', required=True, default=None)
+def state(state, port):
+    """Configure the muxcable to {active/standby} hwmode mode"""
```
Comment (jleveque) [inline config/muxcable.py:279]: Suggest rewording: `Configure the muxcable mux state` or similar

File: show/muxcable.py [hunk 22]
```diff
@@ -412,3 +432,214 @@ def cableinfo(port):
 
     body = [[vendor, part_num]]
     click.echo(tabulate(body, headers=headers))
+
+
+@muxcable.group(cls=clicommon.AbbreviationGroup)
+def hwmode():
+    """start configuring cable in hardware mode on a port"""
```
Comment (jleveque) [inline show/muxcable.py:439]: This is a `show` command, which does not configuration. Please rephrase this appropriately.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: config/muxcable.py (hunk 20 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: config/muxcable.py (hunk 21 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: show/muxcable.py (hunk 22 above)
Comment (vdahiya12) [reply show/muxcable.py]: fixed

File: config/muxcable.py [hunk 23]
```diff
@@ -257,3 +264,208 @@ def disable(port, target):
...
+
+
+@hwmode.command()
+@click.argument('state', metavar='<operation_status>', required=True, type=click.Choice(["active", "standby"]))
+@click.argument('port', metavar='<port_name>', required=True, default=None)
+def state(state, port):
+    """Configure the muxcable mux state {active/standby}"""
+
+    per_npu_statedb = {}
+    transeiver_table_keys = {}
```
Comment (jleveque) [inline config/muxcable.py:282]: Typo: `transceiver_table_keys`

**Review by jleveque — CHANGES_REQUESTED** [review]

File: config/muxcable.py (hunk 23 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

**Review by jleveque — APPROVED** [review]

**Comment by daall** [comment]
@vdahiya12 there is a merge conflict when cherry-picking, please open a PR against the 202012 branch
